src/logic/equiv/fin.lean

```diff
@@ -424,6 +425,56 @@ def fin_prod_fin_equiv : fin m × fin n ≃ fin (m * n) :=
         ... = y.1 : nat.mod_eq_of_lt y.2),
   right_inv := λ x, fin.eq_of_veq $ nat.mod_add_div _ _ }
 
+/-- The equivalence between `a` and `(a / n, a % n)` for nonzero `n`.
+
+This is like `fin_prod_fin_equiv.symm` but with `m` infinite. -/
+@[simps]
+def nat.div_mod_equiv (n : ℕ) [ne_zero n] : ℕ ≃ ℕ × fin n :=
+{ to_fun := λ a, (a / n, fin.of_nat' a),
+  inv_fun := λ p, p.1 * n + p.2,  -- TODO: is there a canonical order of `*` and `+` here?
+  left_inv := λ a, nat.div_add_mod' _ _,
+  right_inv := λ p, begin
+    refine prod.ext _ (fin.ext $ nat.mul_add_mod_of_lt p.2.is_lt),
+    dsimp only,
+    rw [add_comm, nat.add_mul_div_right _ _ (ne_zero.pos n), nat.div_eq_of_lt p.2.is_lt, zero_add],
+  end }
+
+/-- The equivalence between `a` and `(a / n, a % n)` for nonzero `n`. -/
+@[simps]
+def int.div_mod_equiv (n : ℕ) [ne_zero n] : ℤ ≃ ℤ × fin n :=
+{ to_fun := λ a, (a / n, (a.nat_mod n)),
+  inv_fun := λ p, p.1 * n + (p.2 : ℕ),
+  left_inv := λ a, begin
+    dsimp only,
+    convert int.div_add_mod' _ (n : ℤ),
+    rw fin.coe_coe_of_lt ,
+    { rw [int.nat_mod, int.to_nat_of_nonneg],
+      refine int.mod_nonneg _ (nat.cast_ne_zero.mpr $ ne_zero.ne n), },
+    rw ←int.coe_nat_lt,
+    { rw [int.nat_mod,
+        int.to_nat_of_nonneg (int.mod_nonneg _ (nat.cast_ne_zero.mpr $ ne_zero.ne n))],
+      refine int.mod_lt_of_pos _ _,
+      rw [←int.coe_nat_zero, int.coe_nat_lt],
+      exact ne_zero.pos n, },
```

Comment:
Really we just want this lemma, but it should go in `data.int.cast.lemmas`, and we're not importing that here, but it's short so you can just inline it:
```lean
@[simp, norm_cast]
lemma int.coe_nat_mod_coe (a : ℤ) (n : ℕ) (hn : n ≠ 0) : (a.nat_mod n : ℤ) = a % n :=
int.to_nat_of_nonneg (int.mod_nonneg a (by exact_mod_cast hn : (n : ℤ) ≠ 0))
```
and inlined it becomes:
```suggestion
    have key : (a.nat_mod n : ℤ) = a % n := int.to_nat_of_nonneg
      (a.mod_nonneg (by exact_mod_cast (ne_zero.out : n ≠ 0) : (n : ℤ) ≠ 0)),
    simp [key],
```

Reply:
We can just import `data.int.cast.lemmas`. `logic.equiv.fin` is not such a basic file (look at the files importing it).

Reply:
This is now moot I think, with the new golf.